File: src/main/java/com/evergreenclient/hudmod/utils/Position.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 */

package com.evergreenclient.hudmod.utils;

import net.minecraft.client.gui.ScaledResolution;

public class Position {

    private float x, y, scale;

    Position(float x, float y, float scale) {
        this.x = x;
        this.y = y;
        this.scale = scale;
    }

    public static Position getPositionWithRawPositioning(int x, int y, float scale, ScaledResolution resolution) {
        return new Position(MathUtils.getPercent(x, 0, resolution.getScaledWidth()), MathUtils.getPercent(y, 0, resolution.getScaledHeight()), scale);
    }

    public static Position getPositionWithScaledPositioning(float x, float y, float scale) {
        return new Position(x, y, scale);
    }

    public int getRawX(ScaledResolution resolution) {
        return (int) (resolution.getScaledWidth() * x);
    }

    public int getRawY(ScaledResolution resolution) {
        return (int) (resolution.getScaledHeight() * y);
    }

    public float getXScaled() {
        return x;
    }

    public float getYScaled() {
        return y;
    }

    public void setRawX(int x, ScaledResolution resolution) {
        this.x = MathUtils.getPercent(x, 0, resolution.getScaledWidth());
    }

    public void setRawY(int y, ScaledResolution resolution) {
        this.y = MathUtils.getPercent(y, 0, resolution.getScaledHeight());
    }

    public void setScaledX(float x) {
        this.x = x;
    }

    public void setScaledY(float y) {
        this.y = y;
    }

    public float getScale() {
        return scale;
    }

    public void setScale(float scale) {
        this.scale = scale;
    }

    public void add(Position b) {
        x += b.x;
        y += b.y;
        scale += b.scale;
    }

    public void subtract(Position b) {
        x -= b.x;
        y -= b.y;
        scale -= b.scale;
    }

    public static Position center() {
        return new Position(0, 0, 1);
    }

}

File: core/src/main/kotlin/dev/isxander/evergreenhud/utils/Position.kt
/*
 | EvergreenHUD - A mod to improve on your heads-up-display.
 | Copyright (C) isXander [2019 - 2021]
 |
 | This program comes with ABSOLUTELY NO WARRANTY
 | This is free software, and you are welcome to redistribute it
 | under the certain conditions that can be found here
 | https://www.gnu.org/licenses/gpl-3.0.en.html
 |
 | If you have any questions or concerns, please create
 | an issue on the github page that can be found here
 | https://github.com/isXander/EvergreenHUD
 |
 | If you have a private concern, please contact
 | isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.utils

import dev.isxander.evergreenhud.compatibility.universal.RESOLUTION

class Position private constructor(var scaledX: Float, var scaledY: Float, var scale: Float) {

    var rawX: Float
        get() = RESOLUTION.scaledWidth.toFloat() * scaledX
        set(x) { scaledX = MathUtils.getPercent(x, 0f, RESOLUTION.scaledWidth.toFloat()) }
    var rawY: Float
        get() = RESOLUTION.scaledHeight.toFloat() * scaledY
        set(y) { scaledY = MathUtils.getPercent(y, 0f, RESOLUTION.scaledHeight.toFloat()) }

    companion object {
        fun rawPositioning(x: Int, y: Int, scale: Float = 1f): Position =
            Position(MathUtils.getPercent(x.toFloat(), 0f, RESOLUTION.scaledWidth.toFloat()), MathUtils.getPercent(y.toFloat(), 0f, RESOLUTION.scaledHeight.toFloat()), scale)

        fun scaledPositioning(x: Float, y: Float, scale: Float = 1f): Position =
            Position(x, y, scale)
    }

}